17419 Date range picker in place editor

knight-bubble opened the pull request:
[`* `#18939` Date picker inplace editor`](http://community.openproject.org/work_packages/18939)

Commit:
Added date range picker on details pane.

Commit:
Fix updating of date range in initial spans.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://hirtie-maxim-feature-17419-date-picker-in-place-editor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://hirtie-maxim-feature-17419-date-picker-in-place-editor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
Fix missed semicolon and line length.

ulferts commented:
@hirtie-maxim I tried the element and I like the behaviour but there seems to be one limitation/bug in that I cannot switch month in order to set a range that spans more than one month.

When I switch the month, the start date is switched to be in that month as well. If it is not possible to work around that limitation, the UI is not suitable.

Commit:
Add of manual editing of date in input.

Commit:
Add semicolon to the end of line.

Commit:
Add tests for date range picker.

Commit:
Fix date range picker spec.

Commit:
Fix tests.

Commit:
Fix positioning in Firefox

0xF013 commented on frontend/public/templates/components/inplace_editor/display/daterange.html:
```diff
@@ -0,0 +1 @@
+<op-date date-value="dataObject.startDate"></op-date> - <op-date date-value="dataObject.dueDate"></op-date>
```
@hirtie-maxim linebreak please

ulferts commented:
@hirtie-maxim not looking at the code in my review I noticed the following:
- cannot change first day of the week (Sunday, Saturday, Monday). This should be configurable via the setting in "Modules" -> "Administration" -> "Settings" -> "Display"
- cannot set the values to null. Once you have start/end date set to anything, you can not choose to set them to null again.
- When no date is set, the value is displayed as "Invalid date - invalid date". This should read "No start date set - No end date set". 
- I can always only set ranges, not individual dates for example when I only want to specify the start or end date. This might break the idea of having one field with a date range picker.
- The calendar week is not displayed.
- I checked with PM and the requirement is to always display the dates in the format YYYY/MM/DD in while in edit mode. When in show mode, the display should be according to the settings in the admin section. Currently this admin setting is used for both which I personally find more desirable.

Commit:
Fix date range picker.

Add button for clearing dates. Now is possible to set one date.
Add week days in calendar and the starting of week is possible to set
from settings.

Commit:
Changes to range datepicker.

Commit:
Refactor of date range picker added new use cases of it.

floriank commented:
@hirtie-maxim there is still a breaking test in the `spec` group, which seems to be flickering (or caused by a solar flare)

 I restarted the build, let's wait and see, as I doubt your changes broke it. I'd like you to have a look at the @houndci comments though.

0xF013 commented:
@floriank we can always xit the test

floriank commented:
@0xF013 Yeah maybe we could.
@hirtie-maxim Are these @houndci complaints fixable? I currently cannot check as I am sitting in the ubahn :laughing:

0xF013 commented:
@floriank I was making a joke on my test-handling technique.

floriank commented:
I need an `irony` smiley.

Once again: summoning the presence of @hirtie-maxim - what about hound? I reviewed that PR and it looks good so far - I'd only like the code style updated accordingly.

Commit:
Add ignore file for hound.

Commit:
Fix hound issues.

knight-bubble commented:
@floriank, should it be updated for specs? Tried to add ignore comments, or to add the specs to jshintignore file, but without effect.

floriank commented:
Yes, specs and tests should follow the conventions.

0xF013 commented:
@floriank some of protractor instructions are a funny DSL that makes hound go rabid.

floriank commented:
@0xF013 didin't know that! I suppose it's okay then - thanks for the time investment _makesMentalNote_

floriank commented:
the `z-index` on the datepicker widget is currently broken for me in chrome:

![screenshot from 2015-03-25 16 32 09](https://cloud.githubusercontent.com/assets/498241/6828210/ba415f5c-d30c-11e4-83aa-dad4802dd777.png)

upping it shows the widget again:

![screenshot from 2015-03-25 16 32 50](https://cloud.githubusercontent.com/assets/498241/6828215/c1f1077a-d30c-11e4-8a16-02c28ece24cd.png)

Also, I believe the "no <start/end> date set" should be capitalized.

ulferts commented:
The cuke failure is a query_menu_items failure and as such, it can be ignored.

Commit:
Fix tests and z-index of datepicker.

ulferts commented:
I am ignoring the styling flaws as those will be addressed separately. This leaves the following I noticed:

## No ... date

There is "no ... date" displayed in the edit fields when there shouldn't and the value of the date picker is then set to the current date:

![image](https://cloud.githubusercontent.com/assets/617519/6845446/23a2f9c6-d3b5-11e4-9e90-b40019f8ab85.png)

clicking on the field:
![image](https://cloud.githubusercontent.com/assets/617519/6845451/37667de8-d3b5-11e4-8e6f-f1c6d7ba1fe0.png)

And the date picker on the due date:
![image](https://cloud.githubusercontent.com/assets/617519/6845470/5589242e-d3b5-11e4-97a8-2ac94127dbd3.png)

## Placement when opening to the top

When not enough space is available below, the date picker opens to the top (good). But the placement is off then:
![image](https://cloud.githubusercontent.com/assets/617519/6845488/9a7b84a0-d3b5-11e4-94a7-a0f7908778a5.png)

## "No ... date" not placeholder attribute

The "No ... date" should only be a placeholder so that when I start typing I don't have to remove the whole text (is done automatically when using the picker.

## Space between start and end date:

Currently, there is no space between the dash and the start and end date. This makes it hard to read especially if the format is DD-MM-YYYY
![image](https://cloud.githubusercontent.com/assets/617519/6845679/5c32d2d2-d3b7-11e4-9b17-e580dccee1c7.png)

## Error when having format set to anything but DD/MM/YYYY

The format chosen via the administrative settings (Admin -> Settings -> Display) is correctly displayed in read mode. But when trying to update the field then, errors are returned as the request to the server is no longer constructed correctly:
![image](https://cloud.githubusercontent.com/assets/617519/6845799/8514eeaa-d3b8-11e4-9894-ed87818412b8.png)
Console for that:
![image](https://cloud.githubusercontent.com/assets/617519/6845807/9d2409cc-d3b8-11e4-9c7b-30885f3133bb.png)

Or:
![image](https://cloud.githubusercontent.com/assets/617519/6845813/afd2678a-d3b8-11e4-822f-2cd19f157d76.png)
Console for that:
![image](https://cloud.githubusercontent.com/assets/617519/6845818/c56d3c32-d3b8-11e4-986a-262ed0a06332.png)

knight-bubble commented:
@ulferts, You commented:
I checked with PM and the requirement is to always display the dates in the format YYYY/MM/DD in while in edit mode. When in show mode, the display should be according to the settings in the admin section. Currently this admin setting is used for both which I personally find more desirable.

So should be the date format chosen in edit mode?

floriank commented:
@hirtie-maxim I believe @ulferts talked about personal preference here. I concur with that opinion - I would say that the display in edit mode, as well as in read mode should follow the configured format in the settings. 

I would also consider this a detail - if it takes too much time to change it so that the edit mode always uses the ISO8601 Format `YYYY-MM-DD` and the read mode uses the configured version, then please do not spend time on this detail.

By the way, @ulferts is also out of commission for 2 months and will probably not respond.

Commit:
Fix showing of date in edit mode.

floriank commented:
@hirtie-maxim Please remove the WIP label once you consider this mergable.

knight-bubble commented:
This pull request will be cherry-picked by @0xF013, once he is done with his huge PR.

floriank commented:
Ah, so it was this one - we were unsure a moment ago.

NobodysNightmare commented:
@hirtie-maxim Please rebase your work onto the dev branch.

Commit:
Rebased on @0xF013's branch

NobodysNightmare commented:
Style issues:
![date-alignment](https://cloud.githubusercontent.com/assets/453584/7071390/3baade40-dee5-11e4-9cbe-1fc476093a60.PNG)

The vertical alignment of the label (**Date**) and the date range picker is off.
- **red**: The baselines do not match (_good_ baselines of other elements also highlighted)
- **pink**: The label is not centered w.r.t. the input fields

NobodysNightmare commented:
There is no edit control for custom fields of type `Date`.

Is this intended?

NobodysNightmare commented:
When scrolling the date picker stays at its absolute position and does not scroll with the pane, while other elements (like the edit controls) scroll with the pane:

![date-picker-scroll](https://cloud.githubusercontent.com/assets/453584/7071551/4987a7b8-dee6-11e4-9f51-aa15e6d51755.PNG)

Commit:
Fix baseline, scrolling for datepicker and tests.

NobodysNightmare commented:
You are now hiding the date picker on scrolling...
This means that after scrolling I have to deselect the input field and select it again to make the date picker visible again.

Isn't it possible to keep it open, but aligned with the scrolling of the details pane (like the edit controls do)?

NobodysNightmare commented:
I think the date picker should be hidden after clicking `Clear`, as it is otherwise hard to submit the changes, because the picker overlaps the edit controls.

Hiding the picker after clicking "clear" would also be consistent with the fact that the picker is hidden after you click on a date.

NobodysNightmare commented:
The travis build is failing because you commited chagnes to the `Gemfile.lock`

NobodysNightmare commented:
We want to merge this earlier, though not "perfect".

I will re-issue that PR with a reverted `Gemfile.lock` and ideally get that PR merged than.

For the remaining issues that arose during this PR I will create separate work packages.

NobodysNightmare commented:
closing in favor of https://github.com/opf/openproject/pull/2816
